### PR TITLE
fix tooltips disappearing after trying to interact during their fade out animation

### LINKED ISSUE #31646: Bootstrap 5 Tooltip with inline SVG
I've been testing the Bootstrap v5 tooltips (https://v5.getbootstrap.com/docs/5.0/components/tooltips/) but having an issue using them with inline SVGs. The tooltip triggers off when hovering over the SVG - resulting in a flicker of the tooltip showing/hiding.
Simplified test-case CodePen here:
https://codepen.io/coliff/pen/JjXpOJY

The same code works fine with Bootstrap v4. 
https://codepen.io/coliff/pen/VwadMKX

(side note; as a stop-gap fix adding the css `svg {pointer-events: none}` fixes the issue).

This issue was discussed in a bit more detail at: https://github.com/twbs/bootstrap/discussions/31618

### LINKED ISSUE #32372: v5: Bootstrap tooltips don't appear after numbers of focusing
- Operating System: **Windows 10**
- Browser: **Chrome version 87.0.4280.88**

Titled problem occurs even on official docs of bootstrap website at [here](https://getbootstrap.com/docs/5.0/components/tooltips). Tooltips all after some hovering are disappeared and not be accessible any more.

### PR DESCRIPTION
**Pull request by RyanBerliner**
Resolves #32372, resolves #31646

This points out that components that transition may have a blind spot in testing. Since transition duration & delay is computed using [getTransitionDurationFromElement(element)](https://github.com/twbs/bootstrap/blob/main/js/src/util/index.js#L76), which gets transitions from css, its always going to be returning 0 in the unit tests unless otherwise mocked or styled inline. That means that any components that use this transition method will have tests that mostly ignore them... which may not actually matter (likely true), but at least in this case it does matter.

Ideal in testing you'd mock the actual `getTransitionDurationFromElement(element)` imo, and just return the duration, however I was having some trouble getting jasmine spys to properly do this. I went for the next best thing in this case, which is mocking computed styles to get a non-zero return from that function. Happy to change it if anyones got ideas.

Preview: https://deploy-preview-33289--twbs-bootstrap.netlify.app/docs/5.0/components/tooltips/

### CODE REVIEW

**Comment by XhmikosR** [comment]
This fixes the issue, but doesn't fix the issue with the tooltip randomly showing up in the wrong position :/

https://user-images.githubusercontent.com/349621/110216413-bd997d80-7eb7-11eb-99f3-dfdfc8ab597b.mp4

It probably is a separate issue, though.

But yeah, feel free to do what you think it's best with mocking since we have been already bitten by this issue.

**Comment by RyanBerliner** [comment]
Yeah - I figured I'd leave that to #31646 as the positioning issues seem to have everything to do with mouseout and mouseover behavior being investigated there already. You can actually very consistently get the position of the tooltip to mess up if you slowly move your mouse onto the button from where the tooltip is going to appear... because it triggers the following events

- mouse hovers over trigger, to start showing tooltip
- the tooltip shows, but in a position that the mouse is on, which is currently registers a "leave" because the mouse isn't actually in trigger anymore. So, the tooltip hides
- with the tooltip hidden, the cursor is back on the trigger... showing the tooltip again... and the cycle continues until positioning fails

https://user-images.githubusercontent.com/22206986/110216758-e2bdca00-7e7e-11eb-98aa-1f62c9f3d6c1.mov

Which is of course why adding `pointer-events:none;` to the tooltip works as mentioned earlier. 

Long story short, that seemed separate from the already being dealt with #31646 so I've ignored it for this. It does though seem that the code included in this pr makes solving the other issue less complicated. I'll leave this pr as is, lmk if you'd like anything changed here or a hand on the other issue

**Comment by XhmikosR** [comment]
If you can help us with the other issue, please go ahead :)

Lack of tests caused these issues in the first place. Currently the tooltip and event handling issues are blocking us.

**Comment by RyanBerliner** [comment]
This should now resolve #31646 as well. My initial thinking was a bit off. The issue with tooltips ending up in weird locations was due to trying to create a NEW tooltip before the other one was destroyed. I also decided NOT to handle tooltip/trigger overlapping, because since it looks like tooltip offset is allowed to be customized by users easily, in the edge cases that it matters users can just make the offset larger.

Lastly, I changed the default ordering of `fallbackPlacements` to be `['top', 'bottom', 'left', 'right']` since it seemed strange to me that if "top" tooltips where scrolled to the top of the viewport they would do anything but go to the bottom.

**Comment by rohit2sharma95** [comment]
Thanks for the PR @RyanBerliner

Change in the `fallbackPlacements` placement seems unrelated here :slightly_smiling_face: 
The default value was added here to override the default value of Popper because Popper does not flip the element if there is not enough space on the opposite side. And it was kept clockwise (starting from the top) as that is the usual pattern (like giving CSS padding and margin). Your opinion is also good but that is just for the top and bottom placement if the default placement is on the right side and the user scrolls horizontally, the tooltip will move to top => bottom => left (that is neither the opposite of right not clockwise :slightly_smiling_face:).

That default pattern may not be preferred by everyone but that can be overridden by the configuration of the tooltip.

Ref: #32843 and https://popper.js.org/docs/v2/modifiers/flip/#fallbackplacements

File: js/src/tooltip.js [hunk 1]
```diff
@@ -283,6 +283,10 @@ class Tooltip extends BaseComponent {
 
     EventHandler.trigger(this._element, this.constructor.Event.INSERTED)
 
+    if (this._popper) {
+      this._popper.destroy()
+    }
+
```
Comment (rohit2sharma95) [inline js/src/tooltip.js:289]: Instead of destroying the previous popper instance and then creating a new one, you can just use the previous instance if it exists :thinking: 

```js
if (!this._popper) {
  this._popper = Popper.createPopper(this._element, tip, this._getPopperConfig(attachment))
}
```

**Review by rohit2sharma95 — COMMENTED** [review]
Can you please add tests for the second issue as well :slightly_smiling_face:

**Comment by RyanBerliner** [comment]
1. `fallbackPlacement` - I've reverted this to the clockwise rotation it was before
2. Testing - I already added a tested for the "strange popper placement" bug. You'll see in [this previous commit I add a check for consistent tooltip placement](https://github.com/RyanBerliner/bootstrap/commit/2d86690a26bb3f7abc68c34280c65f6441ac5f9c). Without the fix in place this test fails identically to how you'd see it in the browser, jumping from 'top' to 'right'. If you'd rather I split it out into its own test, or spy on its implementation more I can do that... not sure how far into the weeds to go testing implementation vs outcome. happy to change it.

File: js/src/tooltip.js (hunk 1 above)
Comment (RyanBerliner) [reply js/src/tooltip.js]: yes, with one caveat, the existing popper needs to be updated in case the the tooltip text has changed (which means it needs to be repositioned). This is applicable to the "copy to clipboard" buttons in the documentation for example. I've applied this change.

**Comment by XhmikosR** [comment]
@RyanBerliner I haven't looked at your changes yet, but I wanted to say a huge thank you for jumping into these issues! ❤️ 

The tooltip and event issues are the ones blocking as AFAICT, so this would unblock beta3 and generally v5.

File: site/content/docs/5.0/components/tooltips.md [hunk 2]
```diff
@@ -261,7 +261,7 @@ Note that for security reasons the `sanitize`, `sanitizeFn`, and `allowList` opt
     <tr>
       <td><code>fallbackPlacements</code></td>
       <td>array</td>
-      <td><code>['top', 'right', 'bottom', 'left']</code></td>
+      <td><code>['top', 'bottom', 'left', 'right']</code></td>
```
Comment (rohit2sharma95) [inline site/content/docs/5.0/components/tooltips.md:264]: Ah, still not reverted :smile:

Comment (RyanBerliner) [reply site/content/docs/5.0/components/tooltips.md:264]: whoops - forgot to stage that file... incoming...

File: js/src/tooltip.js [hunk 3]
```diff
@@ -283,7 +283,11 @@ class Tooltip extends BaseComponent {
 
     EventHandler.trigger(this._element, this.constructor.Event.INSERTED)
```
Comment (RyanBerliner) [inline js/src/tooltip.js:284]: Also, please consider this event. Depending on how you anticipate this event to be used by people, it may make most sense to only trigger it on true, _new popper_ creation, otherwise it will be triggered every time someone re-enters a tooltip trigger while its transitioning to hidden.

Comment (rohit2sharma95) [reply js/src/tooltip.js:284]: True, it should be triggered only when the tooltip is inserted in the DOM.

**Comment by XhmikosR** [comment]
@rohit2sharma95 @GeoSot LGTY? I think it's a clean way to fix the issues we are hitting.

File: js/src/tooltip.js [hunk 4]
```diff
@@ -646,7 +653,7 @@ class Tooltip extends BaseComponent {
     if (event) {
       context._activeTrigger[
         event.type === 'focusout' ? TRIGGER_FOCUS : TRIGGER_HOVER
-      ] = false
+      ] = context._element.contains(event.relatedTarget)
```
Comment (rohit2sharma95) [inline js/src/tooltip.js:656]: This change is not required IMO :thinking:
Since the active trigger should be false when the tooltip is about to be hidden.

Comment (alpadev) [reply js/src/tooltip.js:656]: @rohit2sharma95 I think this is neccessary to make it work with SVG and nested elements that cause mouseover to trigger

Comment (rohit2sharma95) [reply js/src/tooltip.js:656]: IMO, now this should be handled by:
```js
if (this._isWithActiveTrigger()) {
  return
}
```

Comment (alpadev) [reply js/src/tooltip.js:656]: Mh maybe you're correct then. Another reason could be that this also prevents `hide.bs.tooltip` event from getting triggered but now idea if that been the reason.

Comment (GeoSot) [reply js/src/tooltip.js:656]: I agree with Rohit. 
```js

if (this._isWithActiveTrigger()) {
  return
}
```

seems engough

**Comment by GeoSot** [comment]
Maybe I am wrong and maybe it not the proper place to ask for this, but I think
`_activeTrigger`  ""reason" is not used somewhere except line 198. 
```js
context._activeTrigger.click = !context._activeTrigger.click
```
 
Wouldn't be easier to keep only  `_activeTrigger` as boolean value?
I see much 'noise' without purpose

File: js/src/tooltip.js (hunk 4 above)
Comment (alpadev) [reply js/src/tooltip.js:656]: I'm not sure either why exactly it was decided to do it that way but I tried to adapt the behaviour to their native counterparts on the `EventHandler` module.
During my tests on the tooltip page tho I realized there is a problem with the tooltip plugin itself so I rather focused on this problem instead.
Like you, I found the reason behind the tooltip from jumping around to be within the initialization of popper. But I didn't realize the problem with the transition handlers until later. I partly did some clean-up (I started by uncommenting `_hoverState` as someone pointed out to be a problem and eventually got rid of it) and after reading and understanding the code better I eventually realized where the problem comes from but I still had to come up with a proper solution for this.
During testing I also realized the same problem with the element loosing the hover state with the tooltip showing up so I got here to ask if this is something to be fixed as well. Thats when I saw your pull request and well looks like finishing the `EventHandler` would have been a better time investment.
Also I found another problem with the `noop` handlers for the IOS fix - they are not getting removed properly but no idea if this will have an impact on performance.

**Comment by alpadev** [comment]
> Maybe I am wrong and maybe it not the proper place to ask for this, but I think
> `_activeTrigger` ""reason" is not used somewhere except line 198.
> 
> ```js
> context._activeTrigger.click = !context._activeTrigger.click
> ```
> 
> Wouldn't be easier to keep only `_activeTrigger` as boolean value?
> I see much 'noise' without purpose

While I think this could be refactored to be a boolean it also keeps track if something is focused while not being hovered for example.

File: js/src/tooltip.js (hunk 4 above)
Comment (RyanBerliner) [reply js/src/tooltip.js:656]: I've [added a test in this commit](https://github.com/twbs/bootstrap/pull/33289/commits/122a4af76c83ff9c60f616cb69dec9e0d146ce62) which show why it important this change is important.

**Comment by alpadev** [comment]
@RyanBerliner added a draft PR https://github.com/twbs/bootstrap/pull/33310 that would address what you mentioned earlier (some little revenge for you being faster with this one 😁)

File: js/tests/unit/tooltip.spec.js [hunk 5]
```diff
@@ -708,6 +708,100 @@ describe('Tooltip', () => {
...
+        const moveMouseToChildEvent = createEvent('mouseout')
+        Object.defineProperty(moveMouseToChildEvent, 'relatedTarget', {
+          value: triggerChild
+        })
+
+        tooltipEl.dispatchEvent(moveMouseToChildEvent)
+      })
+
+      tooltipEl.addEventListener('mouseout', () => {
+        expect(tooltip.hide).toHaveBeenCalledTimes(0)
```
Comment (rohit2sharma95) [inline js/tests/unit/tooltip.spec.js:735]: Better :thinking:  => `expect(tooltip.hide).not.toHaveBeenCalled()`

File: js/tests/unit/tooltip.spec.js [hunk 6]
```diff
@@ -708,6 +708,100 @@ describe('Tooltip', () => {
...
+
+      const tooltipEl = fixtureEl.querySelector('a')
+      const tooltip = new Tooltip(tooltipEl)
+
+      spyOn(window, 'getComputedStyle').and.returnValue({
+        transitionDuration: '0.15s',
+        transitionDelay: '0s'
+      })
+
+      const instertedFunc = jasmine.createSpy()
```
Comment (rohit2sharma95) [inline js/tests/unit/tooltip.spec.js:785]: Typo **insertedFunc**

**Comment by alpadev** [comment]
I think it's not in the scope of this ticket but isn't `this._hoverState` somewhat redundant to `this._activeTrigger.hover`?

File: js/tests/unit/tooltip.spec.js [hunk 7]
```diff
@@ -708,6 +708,100 @@ describe('Tooltip', () => {
...
+          expect(tooltip.getTipElement().classList.contains('show')).toEqual(false)
+          tooltipEl.dispatchEvent(createEvent('mouseover'))
+        }, 100)
+
+        setTimeout(() => {
+          expect(tooltip._popper).not.toBeNull()
+          expect(tooltip.getTipElement().getAttribute('data-popper-placement')).toBe('top')
+          done()
+        }, 200)
+      }, 0)
```
Comment (GeoSot) [inline js/tests/unit/tooltip.spec.js:769]: @RyanBerliner can you leave a comment on top of first time-out explaining its purpose with zero delay? Or replace zero timeout with 1milisec 

I suppose you wrote it like this to give a minimum time to Tooltip to be rendered

Comment (RyanBerliner) [reply js/tests/unit/tooltip.spec.js:769]: tbh, I know it's lame, but the reason I did it this way was to match styling from [this existing tooltip test.](https://github.com/twbs/bootstrap/blob/main/js/tests/unit/tooltip.spec.js#L682) And given that the following is true:
```javascript
setTimout(() => console.log('here 1'), 0)
console.log('here 2')

------ Outputs ------

"here 2"
"here 1"
```
I figured that it's the preferred syntax in this case. I can change it, but if I do it should probably be changed in the other tests as well for consistency?

Comment (GeoSot) [reply js/tests/unit/tooltip.spec.js:769]: Good point. **cc/** @XhmikosR the monster of code consistency, may give us his opinion

Comment (XhmikosR) [reply js/tests/unit/tooltip.spec.js:769]: Would be nice to streamline these but in a separate PR after this. For now, I just want us to merge this and any other beta3 PRs so that we release beta3 next week :)

Comment (alpadev) [reply js/tests/unit/tooltip.spec.js:769]: One thing to mention here. Practically there is no difference between a timeout of 0 or 1, as there is some minimum timeout defined by the runtime that is >0.

**Review by GeoSot — APPROVED** [review]

**Review by rohit2sharma95 — APPROVED** [review]

**Comment by XhmikosR** [comment]
Thanks @RyanBerliner! If you have time, feel free to help us with any other JS-related PRs :)
